frontend/src/widgets/genericbar.js

```diff
@@ -21,73 +21,66 @@ import { Settings } from '../settings';
-    dropdownItems: PropTypes.array,
-    fontSize: PropTypes.number,
-    height: PropTypes.number,
-    hideDropdown: PropTypes.bool,
-    horizontal: PropTypes.bool,
-    padding: PropTypes.object,
-    params: PropTypes.object,
-    percentData: PropTypes.bool,
-    sortOrder: PropTypes.string,
-    title: PropTypes.string,
-    widgetEndpoint: PropTypes.string,
-    xLabel: PropTypes.string,
-    xLabelTooltip: PropTypes.string,
-    yLabel: PropTypes.string,
-    onDeleteClick: PropTypes.func,
-    onEditClick: PropTypes.func
-  }
+const GenericBarWidget = (props) => {
+  const {
+    barWidth = 20,
+    dropdownItems = ['component', 'env', 'metadata.jenkins.job_name'],
+    fontSize,
+    height,
+    hideDropdown,
+    horizontal,
+    padding = {
+      bottom: 30,
+      left: 150,
+      right: 15,
+      top: 20
+    },
+    params = {},
+    percentData,
+    sortOrder = 'descending',
+    title = 'Recent Run Results',
+    widgetEndpoint = 'run-aggregator',
+    xLabel = '',
+    xLabelTooltip,
+    yLabel = '',
+    onDeleteClick,
+    onEditClick,
+  } = props;
 
-  constructor(props) {
-    super(props);
-    this.title = props.title || 'Recent Run Results';
-    this.params = props.params || {};
-    this.getData = this.getData.bind(this);
-    this.onGroupFieldSelect = this.onGroupFieldSelect.bind(this);
-    this.onWeekSelect = this.onWeekSelect.bind(this);
-    this.state = {
-      data: {},
-      barCharts: [],
-      isLoading: true,
-      genericBarError: false,
-    };
-  }
+  const [data, setData] = useState({});
+  const [barCharts, setBarCharts] = useState([]);
+  const [isLoading, setIsLoading] = useState(true);
+  const [genericBarError, setGenericBarError] = useState(false);
 
-  onGroupFieldSelect = (value) => {
-    this.props.params.group_field = value;
-    this.getData();
-  }
-
-  onWeekSelect = (value) => {
-    this.props.params.weeks = value;
-    this.getData();
-  }
-
-  getData() {
-    this.setState({isLoading: true});
-    let widgetEndpoint = this.props.widgetEndpoint || 'run-aggregator';
-    HttpClient.get([Settings.serverUrl, 'widget', widgetEndpoint], this.params)
+  const getData = () => {
+    setIsLoading(true);
+    HttpClient.get([Settings.serverUrl, 'widget', widgetEndpoint], params)
       .then(response => {
         response = HttpClient.handleResponse(response, 'response');
         if (!response.ok) {
           throw Error(response.statusText);
         }
         return response.json();
       })
-      .then(data => this.setState({data: data, isLoading: false}, () => {
-        this.getBarCharts();
-      }))
+      .then(responseData => {
+        setData({...responseData});
+        setIsLoading(false);
+      })
       .catch(error => {
-        this.setState({genericBarError: true});
+        setGenericBarError(true);
         console.log(error);
       });
   }
 
-  getBarStyle = (key) => {
+  const onGroupFieldSelect = (value) => {
+    params.group_field = value;
```

Comment:
Note how I've separated these things into state in #592, since they can initially be passed as parameters but will be modified by the user and should be tracked in state instead of mutating the parameter.